theano/tensor/opt.py

```diff
@@ -2216,10 +2216,64 @@ def local_cast_cast(node):
             not isinstance(x.owner.op, T.Elemwise) or
             not isinstance(x.owner.op.scalar_op, scalar.Cast)):
         return
-    if node.op.scalar_op.o_type == x.owner.op.scalar_op.o_type:
+
+    type1 = x.owner.op.scalar_op.o_type
+    type2 = node.op.scalar_op.o_type
+    base = x.owner.inputs[0]
+
+    if type1 == type2:
         # We don't need to copy over any stack traces here
         return [x]
 
+    if(upcast(base.dtype, type1.dtype)):
+        # Checking for further redundancy. Eg: int8 -> int32 -> int8
+        if(type2.dtype == base.dtype):
+            return x.owner.inputs
+        else:
+            # Apply the second cast only
+            v = node.op(base)
+            return [v]
```

Comment:
Adding v = copy_stack_trace(node.outputs[0], v) gives "ValueError: A local optimizer tried to remove a Variable that is used".
v = copy_stack_trace([node.outputs[0], node.inputs[0]], v) doesn't seem to solve the problem either.
Any ideas?